### PR TITLE
Add first plugins from Meilisearch

### PR DESCRIPTION
**Pull request by mgabelle**
### What changes are being made and why?

Add plugins from Meilisearch :

- DocumentAdd (add a Map value or a URL string value)
- DocumentGet
- Search
- FacetSearch

---

### How the changes have been QAed?

---

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -53,6 +53,8 @@ dependencies {
     // libs included in the final jar
     // TODO remove it after using the GitHub template as your plugin may not need it
     api "com.google.code.gson:gson:2.11.0"
```
Comment (loicmathieu) [inline build.gradle:55]:
```suggestion
```

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+@Plugin(
+    examples = {
+        @io.kestra.core.models.annotations.Example(
+            title = "Add Document to Meilisearch",
+            code = {
+                """
+                    document: \\"document to be added in JSON\\",
+                    url: \\"url of the meilisearch server\\",
+                    key: \\"masterKey of the meilisearch server\\",
+                    index: \\"index where we want to add the document\\"
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:35]:
```suggestion
                """
                    document: "document to be added in JSON"
                    URL:  "url of the meilisearch server"
                    key:  "masterKey of the meilisearch server"
                    index: "index where we want to add the document"
```

Java text block allow to use double-quotes inside them.
Moreover, we usually provide a "working example" not an example with description of the properties.

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            searchIndex.addDocuments(runContext.render(from));
+
+            return Output.builder()
+                .outputMessage("Document successfully added to index " + runContext.render(index))
+                .build();
+
+        } catch (MeilisearchException e) {
+            logger.debug(e.getMessage());
+            throw e;
+        }
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:73]: The try/catch is not needed if you didn't do anything specific about it as we will already log it

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    private String index;
+
+    @PluginProperty(dynamic = true)
+    private String facetName;
+
+    @PluginProperty(dynamic = true)
+    private String facetQuery;
+
+    @PluginProperty(dynamic = true)
+    private String[] filters;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:55]: Use  List<String> itself, moreover don't forget to render it!

**Review by loicmathieu — CHANGES_REQUESTED** [review]
DocumentGet return and object but DocumentAdd take a file, maybe DocumentGte should be a file?

When searching, you get objects ... should'n'y it be Maps instead?

File: src/main/java/io/kestra/plugin/meilisearch/AbstractMeilisearchConnection.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@SuperBuilder
+@ToString
+@EqualsAndHashCode
+@Getter
+@NoArgsConstructor
+public abstract class AbstractMeilisearchConnection extends Task implements MeilisearchConnectionInterface {
+    private String url;
+    private String key;
+
+    public MeilisearchFactory meilisearchFactory(RunContext runContext) throws IllegalVariableEvaluationException {
```
Comment (fhussonnois) [inline src/main/java/io/kestra/plugin/meilisearch/AbstractMeilisearchConnection.java:20]: It seems that you always use the same pattern, in the child classes, i.e., : 

```java
        MeilisearchFactory factory = this.meilisearchFactory(runContext);
        Client client = factory.getMeilisearchClient();
```

Could be more simple to just create a static factory method `createClient(RunContext)` that directly returns the `Client`.  Creating an internal `MeilisearchFactory` class doesn't seem relevant here.

**Review by fhussonnois — COMMENTED** [review]
Minor comment, for commit messages we follow this convention: https://www.conventionalcommits.org/

File: build.gradle (hunk 1 above)
Comment (mgabelle) [reply build.gradle:55]: Thanks for catching this. 
I am wondering should I store DocumentGet in a ion file or return it like that ? 
Also kind of the same question with DocumentAdd, should I get it from a "json string" or should I except a file link ?

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java (hunk 4 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:55]: Changed to List<String> and use render

File: src/main/java/io/kestra/plugin/meilisearch/AbstractMeilisearchConnection.java (hunk 5 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/AbstractMeilisearchConnection.java:20]: Sure you're right

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java (hunk 3 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:73]: Removed

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            .build();
+
+        ArrayList<HashMap<String, Object>> hits = searchIndex.facetSearch(fsr).getFacetHits();
+
+        File tempFile = runContext.workingDir().createTempFile(".ion").toFile();
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(tempFile))) {
+            oos.writeObject(Optional.of(hits).orElse(new ArrayList<>()));
+        }
+
+        double hitSize = 0.0;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:76]: It's strange that the number of hit is a double, can you match half a document?
Better to be an int or a long (depending on what is supported by Meilisearch).

File: build.gradle [hunk 7]
```diff
@@ -53,6 +53,8 @@ dependencies {
...
+    implementation 'com.meilisearch.sdk:meilisearch-java:0.14.0'
+    implementation 'org.json:json:20211205'
```
[the remark this answers is not in the document]
Comment (loicmathieu) [reply build.gradle:55]: Can you remove the json library if not used?

**Review by loicmathieu — CHANGES_REQUESTED** [review]

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java (hunk 2 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java]: Added explicit flows

File: build.gradle (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgabelle) [reply build.gradle]: done

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java (hunk 6 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java]: Yes, the api return a Double..that's weird..
Converted to Long

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            }))
+            .reduce(Integer::sum)
+            .blockOptional()
+            .orElse(0);
+
+        runContext.metric(Counter.of("documentAdded", count));
+
+        return Output.builder()
+            .outputMessage("Document successfully added to index " + this.index.as(runContext, String.class))
+            .build();
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:94]: Outputs are designed to be used by other tasks.
Here it looks like a log so I'd rather log that at info level (via the runnerCOntext.logger), and use a VoidOutput and return null.

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        )
+    }
+)
+public class DocumentAdd extends AbstractMeilisearchConnection implements RunnableTask<DocumentAdd.Output> {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    @Schema(
+        title = "The file to upload"
+    )
+    @PluginProperty(dynamic = true)
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:68]: `@PluginProperty` are not needed when using `Property` based attributes, please remove them everywhere.

File: src/main/java/io/kestra/plugin/meilisearch/DocumentGet.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Builder
+    @Getter
+    public static class Output implements io.kestra.core.models.tasks.Output {
+        @Schema(
+            title = "JSON Document",
+            description = "Returned document with id"
+        )
+        private final Map document;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/DocumentGet.java:81]:
```suggestion
        private final Map<String, Object> document;
```

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            .facetName(facetName.as(runContext, String.class))
+            .facetQuery(facetQuery.as(runContext, String.class))
+            .filter(runContext.render(filters).toArray(new String[]{}))
+            .build();
+
+        var results = searchIndex.facetSearch(fsr);
+
+        File tempFile = runContext.workingDir().createTempFile(".ion").toFile();
+        try (var output = new BufferedWriter(new FileWriter(tempFile), FileSerde.BUFFER_SIZE)) {
+            Flux<FacetSearchable> hitFlux = Flux.just(results);
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:104]:
```suggestion
            Flux<FacetSearchable> hitFlux = Flux.fromIterable(results);
```

Flux.just() will emit a single element, so if you have a list you want to emit multiple elements so each element will be serialized as a separate object so each element will be in its own line
Here, you serialize all elements into a big array into a single line.

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Builder
+    @Getter
+    public static class Output implements io.kestra.core.models.tasks.Output {
+        @Schema(
+            title = "FacetSearch results"
+        )
+        private final URI uri;
+        private final Long totalHits;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:121]: Please add documentation via `@Schema`

File: src/main/java/io/kestra/plugin/meilisearch/Search.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    @PluginProperty(dynamic = true)
+    private Property<String> index;
+
+    @Override
+    public Search.Output run(RunContext runContext) throws Exception {
+        Client client = this.createClient(runContext);
+        Index searchIndex = client.index(index.as(runContext, String.class));
+        SearchResult results = searchIndex.search(query.as(runContext, String.class));
+        ArrayList<HashMap<String, Object>> hits = results.getHits();
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/Search.java:72]:
```suggestion
        List<HashMap<String, Object>> hits = results.getHits();
```

Best practice: code against interface

File: src/main/java/io/kestra/plugin/meilisearch/Search.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Builder
+    @Getter
+    public static class Output implements io.kestra.core.models.tasks.Output {
+        @Schema(
+            title = "Add document is successful or not",
+            description = "Describe if the add of the document was successful or not and the reason"
+        )
+        private final URI uri;
+        private final Long totalHits;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/meilisearch/Search.java:95]: Please add documentaiton via `@Schema`

**Review by loicmathieu — CHANGES_REQUESTED** [review]

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java (hunk 11 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:104]: Actually the result returned by facetSearch is always one JSON document like below :  
That's why I just put `Flux.just`. But I will rename `results` to `result`.

Example: 
`{
  "facetHits":[
    {
      "value":"fiction",
      "count":7
    }
  ],
  "facetQuery":"fiction",
  "processingTimeMs":0
}`

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java (hunk 8 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:94]: Use log and remove Output to VoidOutput

File: src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java (hunk 9 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/DocumentAdd.java:68]: Removed

File: src/main/java/io/kestra/plugin/meilisearch/DocumentGet.java (hunk 10 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/DocumentGet.java:81]: Changed to `Map<String, Object>`

File: src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java (hunk 12 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/FacetSearch.java:121]: Added docs

File: src/main/java/io/kestra/plugin/meilisearch/Search.java (hunk 14 above)
Comment (mgabelle) [reply src/main/java/io/kestra/plugin/meilisearch/Search.java:95]: Added docs

**Review by loicmathieu — APPROVED** [review]
LGTM

**Pull request merged**